core/src/ast/utils.rs

```diff
@@ -21,3 +24,224 @@ pub fn get_doc_lines(attrs: &[Attribute]) -> String {
 
     lines
 }
+
+#[derive(Clone, PartialEq, Eq, Hash, Serialize, Deserialize, Debug)]
+pub struct RustLink {
+    path: Path,
+    typ: DocType,
+}
+
+#[derive(Clone, PartialEq, Eq, Hash, Serialize, Deserialize, Debug)]
+enum DocType {
+    Struct,
+    StructField,
+    Enum,
+    EnumVariant,
+    EnumVariantField,
+    Trait,
+    FnInStruct,
+    FnInEnum,
+    FnInTrait,
+    DefaultFnInTrait,
+    Fn,
+    Mod,
+    Constant,
+    Macro,
+}
+
+pub fn get_rust_link(attrs: &[Attribute]) -> Option<RustLink> {
+    for attr in attrs {
+        if attr.path.to_token_stream().to_string() == "diplomat :: rust_link" {
+            if let Ok(syn::Meta::List(syn::MetaList { nested, .. })) = attr.parse_meta() {
+                if nested.len() == 2 {
+                    if let (
+                        syn::NestedMeta::Meta(syn::Meta::Path(path)),
+                        syn::NestedMeta::Meta(syn::Meta::Path(typ)),
+                    ) = (&nested[0], &nested[1])
+                    {
+                        if let Some(typ) = typ.get_ident() {
+                            return Some(RustLink {
+                                path: Path::from_syn(path),
+                                typ: match typ.to_token_stream().to_string().as_str() {
```

Comment:
thought: we _could_ support the same syntax rustdoc uses for intra-doc links, with `struct@foo::bar::Baz`, etc. But might not be worth it

Reply:
Does that syntax support all 14 variants of `DocType`? It also seems to be fairly legacy

Reply:
How is that syntax legacy?

It's _supposed_ to support all variants but I think it has a couple missing that should be added upstream too.

Reply:
It's here https://github.com/rust-lang/rust/blob/master/src/librustdoc/passes/collect_intra_doc_links.rs#L1734.

Notice how that doesn't have `FnInX` which is presumably resolved somewhere else, but I don't want to rebuild that whole logic here.

Reply:
> It's not required in docs anymore, so I don't think many people will know it (I didn't). Can you point me to the upstream code?

It's never been _required_ in docs, it's only needed for disambiguation, and really it's more expansive than it needs to be (from rustdoc's POV, it only needs to know the import namespace: type, value, macro, or module).

(I wrote the rustdoc code for all of this)

But anyway, like I said, I'm not sure there's a huge benefit to supporting this syntax.